library/core/src/array/mod.rs

```diff
@@ -393,37 +395,125 @@ impl<T, const N: usize> [T; N] {
         }
         // FIXME: Convert to crate::mem::transmute once it works with generics.
         // unsafe { crate::mem::transmute::<[MaybeUninit<U>; N], [U; N]>(dst) }
-        crate::mem::forget(guard);
+        mem::forget(guard);
         // SAFETY: At this point we've properly initialized the whole array
         // and we just need to cast it to the correct type.
-        unsafe { crate::mem::transmute_copy::<_, [U; N]>(&dst) }
+        unsafe { mem::transmute_copy::<_, [U; N]>(&dst) }
+    }
+}
+
+/// Creates an array `[T; N]` where each array element `T` is returned by the `cb` call.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(build_array)]
+///
+/// let array: [usize; 5] = core::array::build_array(|i| i);
+/// assert_eq!(array, [0, 1, 2, 3, 4]);
+/// ```
+#[inline]
+#[unstable(feature = "build_array", issue = "none")]
+pub fn build_array<F, T, const N: usize>(mut cb: F) -> [T; N]
+where
+    F: FnMut(usize) -> T,
+{
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> = Guard::new(MaybeUninit::first_ptr_mut(&mut array));
+    for (idx, value) in array.iter_mut().enumerate() {
+        value.write(cb(idx));
+        guard.initialized += 1;
+    }
+    // FIXME: Convert to crate::mem::transmute once it works with generics.
+    mem::forget(guard);
+    // SAFETY: It is guaranteed that the whole array will be filled
+    unsafe { mem::transmute_copy::<_, [T; N]>(&array) }
+}
+
+/// Creates an array `[T; N]` where each fallible array element `T` is returned by the `cb` call.
+/// Unlike `build_array`, where the element creation can't fail, this version will return an error
+/// if any element creation was unsuccessful.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(build_array)]
+///
+/// use core::array::try_build_array;
+///
+/// #[derive(Debug, PartialEq)]
+/// enum SomeError {
+///     Foo
+/// }
+///
+/// let array: Result<[usize; 5], SomeError> = try_build_array(|i| Ok(i));
+/// assert_eq!(array, Ok([0, 1, 2, 3, 4]));
+///
+/// let another_array: Result<[usize; 5], SomeError> = try_build_array(|_| Err(SomeError::Foo));
+/// assert_eq!(another_array, Err(SomeError::Foo));
+/// ```
+
+#[inline]
+#[unstable(feature = "build_array", issue = "none")]
+pub fn try_build_array<E, F, T, const N: usize>(mut cb: F) -> Result<[T; N], E>
+where
+    F: FnMut(usize) -> Result<T, E>,
+{
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> = Guard::new(MaybeUninit::first_ptr_mut(&mut array));
+    for (idx, value) in array.iter_mut().enumerate() {
+        match cb(idx) {
+            Err(e) => {
+                // FIXME: Convert to crate::mem::transmute once it works with generics.
+                mem::forget(guard);
```

Comment:
Ah I'm not sure in this case if we want to forgot previously assigned elements, as the user should be able to see the effects of their creation because it's an `FnMut` closure. I'm not sure what the precedent is no this tho, but I just wanted to point out that this might not be what is expected

Reply:
Good question! My reasoning was: If the array isn't full, then it wouldn't be considered at all.

Now I am not sure what the best approach would be here :)

Reply:
IMO, the unsuccessful creation of a fallible element should discard the entire array as stated on the function signature `try_from_fn(...) -> Result<[T; N], E>`.

Previous assigned elements could be retrieved constructing an array of results (`[Resul<T, E>; N]`) or using a dynamic stack vector, aka, `ArrayVec`.